SCP-3392 UPLC simplifier

michaelpj opened the pull request:
This adds essentially a copy of the PIR inliner for UPLC. This has
pretty good results, with a 5-20% speedup and size decrease for many
programs.

michaelpj commented:
New attempt at https://github.com/input-output-hk/plutus/pull/4365.

bezirg commented on plutus-core/common/PlutusCore/InlineUtils.hs:
```diff
@@ -0,0 +1,9 @@
+import Data.Semigroup (Any (..))
+
+newtype InlineHints name a = InlineHints { shouldInline :: a -> name -> Bool }
+    deriving (Semigroup, Monoid) via (a -> name -> Any)
```
that's cool

bezirg commented on plutus-core/plutus-ir/src/PlutusIR/Analysis/Usages.hs:
```diff
@@ -26,15 +26,11 @@ addUsage n usages =
--- | Get a set of all used @n@s.
+-- | Get a set of @n@s which are used at least once.
 allUsed :: Usages -> Set.Set PLC.Unique
 allUsed usages = Map.keysSet $ Map.filter (> 0) usages
```
isn't the filter inside `allUsed` kind of redundant? We never store on Usages map anything that has less than 0 usages, right? But i guess better be exact/correct.

michaelpj replied:
Yes I guess, but it doesn't hurt :shrug:.

bezirg commented on plutus-core/plutus-ir/src/PlutusIR/Compiler.hs:
```diff
@@ -64,10 +66,10 @@ import PlutusPrelude
   Pass { _name      :: String
        , _shouldRun :: forall m e a.   Compiling m e uni fun a => m Bool
-       , _pass      :: forall m e a b. Compiling m e uni fun a => Term TyName Name uni fun b -> m (Term TyName Name uni fun b)
+       , _pass      :: forall m e a. Compiling m e uni fun a => Term TyName Name uni fun (Provenance a) -> m (Term TyName Name uni fun (Provenance a))
```
oh no, Provenance is back!

michaelpj commented on plutus-core/plutus-ir/src/PlutusIR/Compiler.hs:
```diff
@@ -95,7 +97,7 @@ availablePasses :: [Pass uni fun]
     [ Pass "unwrap cancel"        (onOption coDoSimplifierUnwrapCancel)       (pure . Unwrap.unwrapCancel)
     , Pass "beta"                 (onOption coDoSimplifierBeta)               (pure . Beta.beta)
-    , Pass "inline"               (onOption coDoSimplifierInline)             Inline.inline
+    , Pass "inline"               (onOption coDoSimplifierInline)             (\t -> do { hints <- asks (view (ccOpts . coInlineHints)); Inline.inline hints t })
```
Why is `asks` redundant?

I think it's nicer to keep the general pattern of "config args come first, then the 'main' argument", the glue code isn't so bad.

bezirg replied:
> Why is `asks` redundant?
> 
`asks view` === `view` because view works over monadreader 
```
view :: MonadReader s m => Getting a s a -> m a
```

michaelpj replied:
Wow, I hate it.

bezirg replied:
lol

michaelpj commented on plutus-core/plutus-ir/src/PlutusIR/Compiler/Datatype.hs:
```diff
@@ -308,7 +308,9 @@ mkConstructorType :: Datatype TyName Name uni fun (Provenance a) -> VarDecl TyNa
 -- see note [Abstract data types]
 -- FIXME: normalize constructors also here
-mkConstructorType (Datatype _ _ tvs _ _) constr = PIR.mkIterTyForall  tvs $ _varDeclType constr
+mkConstructorType (Datatype _ _ tvs _ _) constr =
+    let constrTy = PIR.mkIterTyForall tvs $ _varDeclType constr
+    in fmap (\a -> DatatypeComponent ConstructorType a) constrTy
```
Correct, I hate it :D I like naming things!

michaelpj commented on plutus-core/plutus-ir/src/PlutusIR/Compiler/Types.hs:
```diff
@@ -51,16 +52,17 @@ data CompilationOpts = CompilationOpts {
-defaultCompilationOpts :: CompilationOpts
-defaultCompilationOpts = CompilationOpts True False False False 8 True True True False
+defaultCompilationOpts :: CompilationOpts a
+defaultCompilationOpts = CompilationOpts True False False False 12 True True True mempty False
```
That would probably be sensible.

michaelpj commented on plutus-core/plutus-ir/src/PlutusIR/Compiler/Types.hs:
```diff
@@ -51,16 +52,17 @@ data CompilationOpts = CompilationOpts {
-defaultCompilationOpts :: CompilationOpts
-defaultCompilationOpts = CompilationOpts True False False False 8 True True True False
+defaultCompilationOpts :: CompilationOpts a
+defaultCompilationOpts = CompilationOpts True False False False 12 True True True mempty False
```
It might actually be unnecessary now - I needed it before I added the multi-beta thing. But it seems harmless.

michaelpj commented on plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs:
```diff
@@ -291,20 +338,21 @@ maybeAddSubst s n rhs = do
         preInlineUnconditional t = do
             usgs <- asks _usages
-            let termIsUsedOnce = Usages.isUsedOnce n usgs
+            -- 'inlining' terms used 0 times is a cheap way to remove dead code while we're here
+            let termUsedAtMostOnce = Usages.getUsageCount n usgs <= 1
```
Yep. But we don't have dead code for UPLC and I wanted to keep things as consistent as possible, so I did it here too.

michaelpj commented on plutus-tx-plugin/test/Lift/bytestring.plc.golden:
```diff
@@ -2,11 +2,15 @@
-        (lam
-          match_Lift_Spec_WrappedBS_i0
-          [ Lift_Spec_WrappedBS_i2 (con bytestring #68656c6c6f) ]
+      (force
+        (delay
```
See https://github.com/input-output-hk/plutus/pull/4365#discussion_r799327943

kwxm reviewed (approved):
I don't understand the details, but I had a look anyway.  I think I suggested optimising UPLC some time ago but the objection was that we don't have types so it's more difficult to be sure that the optimisations are correct, which is very reasonable and now has me worried a bit. Maybe we should have some long-term aim to have some process to improve our confidence.  What could we do?  

  * Formalise the optimsations?  That'd be hard work. 

  * Have property tests comparing execution results before and after optimisation?  We'd need some way of generating random but sensible untyped PLC programs for that, which is also tricky.  Instead of that could we maybe have some fixed set of test cases like the validation scripts? That'd be weaker but not useless 
  
  *  Could we have some sort of type system/static analysis for UPLC? I like that idea, but it'd also be hard.
  
  * I think you said that the problem with doing certain inlinings in PIR was that you had to inline bits of code where it was difficult to transform the type properly.  [Possibly nonsensical suggestion coming up...] Could we have some weaker type system where we replaced the problematic types with some generic "we don't care about the precise type here" type, or would that be discarding too much information? Maybe that would be similar to the "typed UPLC" idea.
  
Anyway, this all looks good and I'm willing to believe that these particular optimisations are safe.

michaelpj commented on plutus-tx/src/PlutusTx/Lift.hs:
```diff
@@ -70,9 +70,10 @@ safeLift x = do
     compiled <- flip runReaderT ccConfig $ compileTerm lifted
     let erased = UPLC.erase compiled
-    db <- UPLC.deBruijnTerm $ UPLC.simplifyTerm erased
+    db <- UPLC.deBruijnTerm =<< UPLC.simplifyTerm usOpts erased
```
Great question. It was there already :laughing: 

Lifting terms *absolutely* produces stuff that needs simplifying: it just runs the compiler! So really we *should* simplify here otherwise we're leaving lots of garbage for people. But then we're simplifying at runtime.

I'm increasingly uncomfortable with `Lift`, partly for these reasons... :cry:

michaelpj commented on plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Inline.hs:
```diff
@@ -0,0 +1,274 @@
+This pass is essentially a copy of the PIR inliner, and should be KEPT IN SYNC with it.
+It's hard to do this with true abstraction, so we just have to keep two copies reasonably
+similar.
+
```
> We have some confidence in the PIR optimiser because of the types, but they're gone in UPLC; however, if the UPLC optimiser is highly similar to the PIR one, maybe some of the confidence carries over.

Yeah, that's kind of my feeling. It would be even better if they actually shared code, but it's very unclear to me how to do that without trying to construct some very awkward abstraction that covers both PIR terms and UPLC terms...

Good idea about the comment.

michaelpj commented:
>  Instead of that could we maybe have some fixed set of test cases like the validation scripts? That'd be weaker but not useless

Yeah, that would be nice. We could again repurpose the nofib tests for this: have optimized and unoptimized versions and check that they do the same thing.

> Could we have some weaker type system where we replaced the problematic types with some generic "we don't care about the precise type here" type, or would that be discarding too much information?

I jokingly suggested something like that, but from a different angle. If we added `unsafeCoerce`, then we could perform the following transformation:

```
[ { (/\ (a : *) . \ (f : t) . ... ) ty } v ]
-----> 
{ (/\ (a : *) . \ (f : t) . ...  ) [ { unsafeCoerce (t[ty/a]) t } v)] ) ty }
```
that is, we can push `v : t[ty/a]` inside the type abstraction by "re-abstracting" it to have the more abstracted type with `unsafeCoerce`.

Of course, this is risky in that you're subverting the typechecker (as we would be with your suggestion), and also the `unsafeCoerce` calls would then get in the way. Overall it didn't seem worth pursuing...
